ThinBridge/ThinBridge
# Add a registry value to disable ThinBridgeSetting.exe

# Which issue(s) this PR fixes:

N/A

# What this PR does / why we need it:

* Introduce a new registry value `HKLM\SOFTWARE\Policies\ThinBridge\DisableSetting`.
  * `1`:  disable to execute ThinBridgeSetting.exe
  * `0`:  enabled to execute ThinBridgeSetting.exe
  * `default(not specified)`:  enabled to execute ThinBridgeSetting.exe
* Add new policy template files (`Resources\GPO\ThinBridge.admx`, `Resources\GPO\ja-JP\ThinBridge.adml`) to modify `HKLM\SOFTWARE\Policies\ThinBridge\DisableSetting`.
  * We can use it by copying `ThinBridge.admx` to `C:\Windows\PolicyDefinitions` and `ja-JP\ThinBridge.adml to C:\Windows\PolicyDefinitions\ja-JP` in local environment.
  * If in domain, we should copy them to `%systemroot%\Sysvol\domain\Policies\PolicyDefinitions` of a domain controller.
  * "コンピューターの構成" -> "管理用テンプレート" -> "ThinBridge" -> "リダイレクト定義設定の起動禁止"
  * ![image](https://github.com/ThinBridge/ThinBridge/assets/15982708/6aaf0bcf-98a5-401e-89ac-690488855bf9)
* Add a check of `HKLM\SOFTWARE\Policies\ThinBridge\DisableSetting` value.
  * If executing  ThinBridgeSetting.exe is disabled,  executing ThinBridgeSetting.exe is blocked and a warning dialog like below  is displayed.
    * ![image](https://github.com/ThinBridge/ThinBridge/assets/15982708/475bcd12-bd63-4a83-918b-c084a911ab92)

# How to verify the fixed issue:

## The steps to verify:

* Download installer from result of Actions
  * E.g. https://github.com/ThinBridge/ThinBridge/actions/runs/9589934579
* Install ThinBridge
* Execute ThinBridgeSetting.exe
  * [x] Confirm that ThinBridgeSetting.exe is not blocked and starts fine.
* Copy `ThinBridge.admx` to `C:\Windows\PolicyDefinitions` and `ja-JP\ThinBridge.adml` to `C:\Windows\PolicyDefinitions\ja-JP`
* Open group policy editor. (Execute `gpedit.msc`)
* Open "コンピューターの構成" -> "管理用テンプレート" -> "ThinBridge" of group policy.
* Double click "リダイレクト定義設定の起動禁止" and open setting dialog.
* Specify enabled.
* Close dialog by clicking "OK" button
* Open registry editor. (Execute regedit.exe)
* Open a registry value `HKLM\SOFTWARE\Policies\ThinBridge\DisableSetting` of registry editor.
  * [x] Confirm that the value exists. ( Sometimes we need to reboot machine.)
  * [x] Confirm that the value is `1`
* Execute ThinBridgeSetting.exe
  * [x] Confirm that ThinBridgeSetting.exe is blocked and the warning dialog is displayed.
  * ![image](https://github.com/ThinBridge/ThinBridge/assets/15982708/475bcd12-bd63-4a83-918b-c084a911ab92)
* Execute ThinBridge.exe
  * [x] Confirm that ThinBridge.exe is not blocked.
* Open "コンピューターの構成" -> "管理用テンプレート" -> "ThinBridge" of group policy.
* Double click "リダイレクト定義設定の起動禁止" and open setting dialog.
* Specify disabled.
* Close dialog by clicking "OK" button
* Open registry editor. (Execute regedit.exe)
* Open a registry value `HKLM\SOFTWARE\Policies\ThinBridge\DisableSetting` of registry editor.
  * [x] Confirm that the value exists. ( Sometimes we need to reboot machine.)
  * [x] Confirm that the value is `0`
* Execute ThinBridgeSetting.exe
  * [x] Confirm that ThinBridgeSetting.exe is not blocked.
* Execute ThinBridge.exe
  * [x] Confirm that ThinBridge.exe is not blocked.
